### PR TITLE
Agharta/eip1014 create2

### PR DESCRIPTION
**Pull request by soc1c**
reapplying #5, can you make that mergable @dziabko? :)

### CODE REVIEW

File: tests/state_test.go [hunk 1]
```diff
@@ -689,10 +689,41 @@ func TestAllETH(t *testing.T) {
 	skipTests["RevertPrecompiledTouch_storage.json/ConstantinopleFix/0"] = "Bug in Test"
 	skipTests["RevertPrecompiledTouch_storage.json/ConstantinopleFix/3"] = "Bug in Test"
 
+	// EIP 684 Implementations
```
Comment (austinabell) [inline tests/state_test.go]: Remove all of these test skips

File: core/execution.go [hunk 2]
```diff
@@ -293,19 +313,101 @@ func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPric
...
-	//if there's an error we return nothing
-	if err != nil && err != vm.ErrRevert {
-		return nil, address, err
-	}
-	return ret, address, err
+	return ret, *address, err
+
 }
 
+// func exec(env vm.Environment, caller vm.ContractRef, address, codeAddr *common.Address, codeHash common.Hash, input, code []byte, gas, gasPrice, value *big.Int, readOnly bool) (ret []byte, addr common.Address, err error) {
```
Comment (austinabell) [inline core/execution.go]: Remove commented out code if you are not using it

File: core/execution.go [hunk 3]
```diff
@@ -223,62 +221,84 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
...
 		caller.ReturnGas(gas, gasPrice)
 
 		return nil, common.Address{}, errCallCreateDepth
 	}
-
 	if !env.CanTransfer(caller.Address(), value) {
 		caller.ReturnGas(gas, gasPrice)
 
-		return nil, common.Address{}, ValueTransferErr("insufficient funds to transfer value. Req %v, has %v", value, env.Db().GetBalance(caller.Address()))
+		return nil, common.Address{}, ErrInsufficientFunds
```
Comment (austinabell) [inline core/execution.go]: I'm pretty sure the ValueTransferErr is checked explicitly, iirc. So you cannot change this error type as it will affect other logic interactions. Why was this changed?

**Review by austinabell — CHANGES_REQUESTED** [review]
See other PR for other changes to be done relating to Create2, but the logic for Create should not be changed in this PR

File: core/execution.go [hunk 4]
```diff
@@ -293,17 +320,13 @@ func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPric
...
 	}
-
-	// When there are no errors but the maxCodeSize is still exceeded, makes more sense than just failing
+	// Assign err if contract code size exceeds the max while the err is still empty.
 	if maxCodeSizeExceeded && err == nil {
 		err = errMaxCodeSizeExceeded
 	}
 
-	//if there's an error we return nothing
-	if err != nil && err != vm.ErrRevert {
```
Comment (austinabell) [inline core/execution.go:303]: Should keep this and not duplicate logic for both creates ?

File: tests/state_test.go [hunk 5]
```diff
@@ -710,8 +714,8 @@ func TestAllETH(t *testing.T) {
 
 func runETHTests(t *testing.T, fileNames []string, skipTests map[string]string) {
 	unsupportedForkConfigs := map[string]bool{
-		"Constantinople":               true,
-		"ConstantinopleFix":            true,
+		"Constantinople": true,
+		// "ConstantinopleFix":            true,
```
Comment (austinabell) [inline tests/state_test.go]: Remove commented out line because it won't return

File: core/execution.go [hunk 6]
```diff
@@ -223,62 +222,90 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
...
-	// only.
+	// Initialise a new contract and set the code that is to be used by the EVM.
+	// The contract is a scoped environment for this execution context only.
 	contract := vm.NewContract(caller, to, value, gas, gasPrice)
-	contract.SetCallCode(nil, crypto.Keccak256Hash(code), code)
+	contract.SetCallCode(codeAddr, codeHash, code)
 	defer contract.Finalise()
 
-	ret, err = env.Vm().Run(contract, nil, false)
+	ret, err := evm.Run(contract, input, readOnly)
```
Comment (austinabell) [inline core/execution.go]: for consistency change `evm.` to `env.Vm().` and remove the variable above

File: core/execution.go [hunk 7]
```diff
@@ -223,62 +222,90 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
...
+
+	//if there's an error we return nothing
+	if err != nil && err != vm.ErrRevert {
+		return nil, address, err
+	}
+	return ret, address, err
+}
+
+// create creates a new contract using code as deployment code.
+func create(env vm.Environment, caller vm.ContractRef, address, codeAddr *common.Address, codeHash common.Hash, input, code []byte, gas, gasPrice, value *big.Int, readOnly bool) ([]byte, common.Address, error) {
```
Comment (austinabell) [inline core/execution.go]: codeAddr, input, and readOnly are unnecessary (they are always nil or false) and can be removed. 

Can also remove codeHash and just calculate in function, but this doesn't matter as much (readability being main reason)

File: core/execution.go [hunk 8]
```diff
@@ -223,62 +222,90 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
 
 // Create creates a new contract with the given code
 func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPrice, value *big.Int) (ret []byte, address common.Address, err error) {
-	// Depth check execution. Fail if we're trying to execute above the limit.
+	nonce := env.Db().GetNonce(caller.Address())
```
Comment (austinabell) [inline core/execution.go:226]: Small detail but nonce is retrieved twice unnecessarily, once here and once in the create() function. If others are fine with this I am too but logic can be refactored a bit to not have this happen

**Review by austinabell — CHANGES_REQUESTED** [review]
Couple small details and then I'm good with it

File: core/execution.go [hunk 9]
```diff
@@ -223,62 +222,82 @@ func StaticCall(env vm.Environment, caller vm.ContractRef, addr common.Address,
...
 	// be stored due to not enough gas set an error and let it be handled
 	// by the error checking condition below.
 	if err == nil && !maxCodeSizeExceeded {
-		dataGas := big.NewInt(int64(len(ret)))
-		// create data gas
-		dataGas.Mul(dataGas, params.CreateDataGas)
-		if contract.UseGas(dataGas) {
-			env.Db().SetCode(address, ret)
+		createDataGas := big.NewInt(int64(len(ret)))
+		createDataGas.Mul(createDataGas, big.NewInt(200))
```
Comment (austinabell) [inline core/execution.go]: Change the hard coded value to params.CreateDataGas

File: core/execution.go [hunk 10]
```diff
@@ -293,17 +312,18 @@ func Create(env vm.Environment, caller vm.ContractRef, code []byte, gas, gasPric
...
-	// When there are no errors but the maxCodeSize is still exceeded, makes more sense than just failing
+	// Assign err if contract code size exceeds the max while the err is still empty.
 	if maxCodeSizeExceeded && err == nil {
 		err = errMaxCodeSizeExceeded
 	}
 
 	//if there's an error we return nothing
 	if err != nil && err != vm.ErrRevert {
-		return nil, address, err
+		return nil, *address, err
```
Comment (austinabell) [inline core/execution.go]: Why is address used as a pointer in this function by the way? Was there a reason this was changed?

**Review by austinabell — CHANGES_REQUESTED** [review]
Small change and I'll be good to merge this into the branch off development

**Review by austinabell — APPROVED** [review]
lgtm

**Review by steviezhang — APPROVED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Comment by soc1c** [comment]
WTF why did you change the base branch @austinabell

**Comment by soc1c** [comment]
the entire point of this PR is getting this into development!!!

**Comment by austinabell** [comment]
@soc1c The agharta changes haven't even been approved by the community, why would we merge them into development now? There is no functional change what would be the benefit to this? 

agharta/dev was just recently branched off of development so it would be ready to merge in whenever the changes are intended to be introduced (and would cause less conflict with the rename if grouped together)

**Comment by soc1c** [comment]
development is unstable. everything goes into development

community approval is unrelated to our dev work

**Comment by GregTheGreek** [comment]
have we re-megred this onto dev?  @soc1c is correct..

**Comment by austinabell** [comment]
Just did it because that was what I was told to hahah, I changed other agharta PR's base and referenced new PR
